src/Airship/Internal/Route.hs

```diff
@@ -10,51 +13,75 @@ import Airship.Resource
 import Data.Monoid
 import Data.Foldable (foldr')
 import Data.Text (Text)
-import Data.HashMap.Strict (HashMap, insert)
 
 #if __GLASGOW_HASKELL__ < 710
 import           Control.Applicative
 #endif
 import Control.Monad.Identity
-import Control.Monad.Writer (Writer, WriterT (..), execWriter)
+import Control.Monad.Trans.State (State, evalState, modify, put, state)
+import Control.Monad.Writer (Writer, WriterT (..), execWriter, tell)
 import Control.Monad.Writer.Class (MonadWriter)
 
 import Data.String (IsString, fromString)
 
+infixl 4 </>
+infixl 5 />
+infixl 5 </
+infixl 3 #>
+
 -- | 'Route's represent chunks of text used to match over URLs.
 -- You match hardcoded paths with string literals (and the @-XOverloadedStrings@ extension),
--- named variables with the 'var' combinator, and wildcards with 'star'.
-newtype Route = Route { getRoute :: [BoundOrUnbound] } deriving (Show, Monoid)
+-- captured variables with the 'var' combinator, and wildcards with 'star'.
+data Route a =
+  Route {
+      getRoute :: [BoundOrUnbound]
+    -- Allow each 'Route' to consume segments and convert to a value 
+    , routeValue :: State [Text] a
```

Comment:
Note, this was just something quick/dirty. There are quite a few ways to represent this depending on the desired outcome. 

Unrelated, we can/should probably add error handling to this as well. That would make adding something like `varInt` easy to support safely.